File: fsharp/dmmf/Domain.fs
namespace OrderTaking.Domain

open Microsoft.FSharp.Data.UnitSystems.SI.UnitNames
open System.Text.RegularExpressions

// Product code info

type WidgetCode = private WidgetCode of string
module WidgetCode =
    let create code =
        if Regex.IsMatch(code, @"W\d{4}") then
            Ok (WidgetCode code)
        else
            Error "`WidgetCode` must begin with a 'W' and be followed by 4 digits"

    let value (WidgetCode code) = code


type GizmoCode = private GizmoCode of string
module GizmoCode = 
    let create code =
        if Regex.IsMatch(code, @"G\d{3}") then
            Ok (GizmoCode code)
        else
            Error "`GizmoCode` must begin with a 'G' and be followed by 3 digits"

    let value (GizmoCode code) = code


type ProductCode =
    | Widget of WidgetCode
    | Gizmo of GizmoCode


type UnitQuantity = private UnitQuantity of uint32 // between 1 and 1000
module UnitQuantity =
    let create qty =
        if qty < 1u then
            Error "`UnitQuantity` cannot be less than 1"
        else if qty > 1000u then
            Error "`UnitQuantity` cannot be greater than 1000"
        else
            Ok (UnitQuantity qty)
    
    let minimum = UnitQuantity 1u

    let value (UnitQuantity qty) = qty


type KilogramQuantity = private KilogramQuantity of decimal<kilogram>
module KilogramQuantity =
    let create qty =
        if qty < 0.05m then
            Error "`KilogramQuantity` cannot be less than 0.05"
        else if qty > 1000.00m then
            Error "`KilogramQuantity` cannot be more than 1000.00"
        else
            let kgs = qty * 1.0m<kilogram>
            Ok (KilogramQuantity kgs)

    let value (KilogramQuantity qty) = qty

    let minimum = KilogramQuantity 0.05m<kilogram>


type OrderQuantity =
    | Unit of UnitQuantity
    | Kilogram of KilogramQuantity


// Helper -- we'll replace this later
type Undefined = exn


type OrderId = Undefined
type OrderLineId = Undefined
type CustomerId = CustomerId of int

type CustomerInfo = Undefined
type ShippingAddress = Undefined
type BillingAddress = Undefined
type Price = Undefined
type BillingAmount = Undefined

type Order = {
    Id : OrderId
    CustomerInfo : CustomerInfo
    ShippingAddress : ShippingAddress
    BillingAddress : BillingAddress
    OrderLines : OrderLine list
    AmountToBill: BillingAmount
    }

and OrderLine = {
    Id : OrderLineId
    OrderId : OrderId
    ProductCode : ProductCode
    OrderQuantity : OrderQuantity
    Price : Price
    }


type AcknowledgementSent = Undefined
type OrderPlaced = Undefined
type BillableOrderPlaced = Undefined

type PlaceOrderEvents = {
    AcknowledgementSent : AcknowledgementSent
    OrderPlaced : OrderPlaced
    BillableOrderPlaced : BillableOrderPlaced
    }

type PlaceOrderError =
    | ValidationError of ValidationError list

and ValidationError = {
    FieldName : string
    ErrorDescription : string
    }

type UnvalidatedOrderLine = {
    ProductCode : string
    OrderQuantity : int
    }

type UnvalidatedOrder = {
    OrderId : string
    CustomerInfo : string // maybe?
    ShippingAddress : string
    OrderLines : UnvalidatedOrderLine list
    }

type ValidatedOrder = Undefined

type ValidationResponse<'a> = Async<Result<'a,ValidationError list>>

type ValidateOrder =
    UnvalidatedOrder -> ValidationResponse<ValidatedOrder>


type PlaceOrder = UnvalidatedOrder -> Result<PlaceOrderEvents,PlaceOrderError>


type QuoteForm = Undefined
type OrderForm = Undefined

type CategorizedMail =
    | Quote of QuoteForm
    | Order of OrderForm


type ProductCatalog = Undefined
type PricedOrder = Undefined

type CalculatePrices = OrderForm -> ProductCatalog -> PricedOrder

type InvoiceId = InvoiceId of int

type UnpaidInvoice = {
    InvoiceId : InvoiceId
}

type PaidInvoice = {
    InvoiceId : InvoiceId
}

type Invoice =
    | Paid of PaidInvoice
    | Unpaid of UnpaidInvoice

File: rust/src/domain.rs
#![allow(dead_code)] // for our sanity while building things!

use futures::Future;
use regex::Regex;

#[derive(PartialEq, Debug)]
pub struct WidgetCode(String);

impl WidgetCode {
    pub fn create(code: &str) -> Result<WidgetCode, String> {
        let re = Regex::new(r"W\d{4}").expect(r"W\d{4} is a valid regex");
        if re.is_match(code) {
            Ok(WidgetCode(String::from(code)))
        } else {
            Err(String::from(
                "`WidgetCode` must begin with a 'W' and be followed by 4 digits",
            ))
        }
    }

    pub fn value(&self) -> &str {
        &self.0
    }
}

#[derive(PartialEq, Debug)]
pub struct GizmoCode(String);

impl GizmoCode {
    pub fn create(code: &str) -> Result<GizmoCode, String> {
        let re = Regex::new(r"G\d{3}").expect(r"G\d{3} is a valid regex");
        if re.is_match(code) {
            Ok(GizmoCode(String::from(code)))
        } else {
            Err(String::from(
                "`GizmoCode` must begin with a 'G' and be followed by 3 digits",
            ))
        }
    }

    pub fn value(&self) -> &str {
        &self.0
    }
}

pub enum ProductCode {
    Widget(WidgetCode),
    Gizmo(GizmoCode),
}

pub struct UnitQuantity(u32);

impl UnitQuantity {
    pub fn create(qty: u32) -> Result<UnitQuantity, String> {
        match qty {
            0 => Err(String::from("`UnitQuantity` cannot be less than 1")),
            1...1000 => Ok(UnitQuantity(qty)),
            _ => Err(String::from("`UnitQuantity` cannot be greater than 1000")),
        }
    }
}

pub struct KilogramQuantity(f32);

enum OrderQuantity {
    Unit(UnitQuantity),
    Kilogram(KilogramQuantity),
}

// Helper -- we'll replace this later
enum Never {}

type OrderId = Never;
type OrderLineId = Never;
pub struct CustomerId(i32);

type CustomerInfo = Never;
type ShippingAddress = Never;
type BillingAddress = Never;
type Price = Never;
type BillingAmount = Never;

struct OrderLine {
    id: OrderLineId,
    order_id: OrderId,
    product_code: ProductCode,
    order_quantity: OrderQuantity,
    price: Price,
}

struct Order {
    id: OrderId,
    customer_info: CustomerInfo,
    shipping_address: ShippingAddress,
    billing_address: BillingAddress,
    order_lines: Vec<OrderLine>,
    billing_amount: BillingAmount,
}

type AcknowledgementSent = Never;
type OrderPlaced = Never;
type BillableOrderPlaced = Never;

struct PlaceOrderEvents {
    acknowledgement_sent: AcknowledgementSent,
    order_placed: OrderPlaced,
    billable_order_placed: BillableOrderPlaced,
}

struct ValidationError {
    field_name: String,
    error_description: String,
}

enum PlaceOrderError {
    ValidationError(Vec<ValidationError>),
}

struct UnvalidatedOrderLine {
    product_code: String,
    order_quantity: i32,
}

struct UnvalidatedOrder {
    order_id: String,
    customer_info: String, // maybe?
    shipping_address: String,
    order_lines: Vec<UnvalidatedOrderLine>,
}

struct ValidatedOrder {
    order_id: String,
}

// NOTE: this uses `Future`, and is therefore rather... more complicated.
type ValidationResponse<T> = Future<Item = T, Error = ValidationError>;

#[allow(unused_variables)] // until we implement this
fn validate_order(unvalidated: UnvalidatedOrder) -> Box<ValidationResponse<ValidatedOrder>> {
    unimplemented!()
}

#[allow(unused_variables)] // until we implement this
fn place_order(unvalidated: UnvalidatedOrder) -> Result<PlaceOrderEvents, PlaceOrderError> {
    unimplemented!()
}

type VaidateOrder = Fn(UnvalidatedOrder) -> Box<ValidationResponse<ValidatedOrder>>;

type QuoteForm = Never;
type OrderForm = Never;

enum CategorizedMail {
    Quote(QuoteForm),
    Order(OrderForm),
}

type ProductCatalog = Never;
type PricedOrder = Never;

fn calculate_prices(order_form: OrderForm, catalog: ProductCatalog) -> PricedOrder {
    unimplemented!()
}

struct InvoiceId(i32);

struct UnpaidInvoice {
    invoice_id: InvoiceId,
}

struct PaidInvoice {
    invoice_id: InvoiceId,
}

enum Invoice {
    Paid(PaidInvoice),
    Unpaid(UnpaidInvoice),
}
